xbmc/guilib/TextureBase.h

```diff
@@ -0,0 +1,117 @@
+/*
+ *  Copyright (C) 2005-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "guilib/TextureFormats.h"
+
+#include <memory>
+#include <string>
+
+enum class TEXTURE_SCALING
+{
+  LINEAR,
+  NEAREST,
+};
+
+/*!
+\ingroup textures
+\brief Base texture class, which holds the state of the texture, as well as all conversion functions.
+*/
+class CTextureBase
+{
+
+public:
+  CTextureBase() = default;
+  ~CTextureBase() = default;
+
+  bool HasAlpha() const { return m_hasAlpha; }
+  void SetAlpha(bool hasAlpha) { m_hasAlpha = hasAlpha; }
+
+  /*! \brief sets mipmapping. do not use in new code. will be replaced with proper scaling. */
+  void SetMipmapping() { m_mipmapping = true; }
+  /*! \brief return true if we want to mipmap */
+  bool IsMipmapped() const { return m_mipmapping; }
+
+  /*! \brief sets the scaling method. scanout systems might support more than NN/linear. */
+  void SetScalingMethod(TEXTURE_SCALING scalingMethod) { m_scalingMethod = scalingMethod; }
+  /*! \brief returns the scaling method. */
+  TEXTURE_SCALING GetScalingMethod() const { return m_scalingMethod; }
+
+  int32_t GetOrientation() const { return m_orientation; }
+  void SetOrientation(int orientation) { m_orientation = orientation; }
+
+  /*! \brief retains a shadow copy of the texture on the CPU side. */
+  void SetCacheMemory(bool bCacheMemory) { m_bCacheMemory = bCacheMemory; }
+  /*! \brief returns true if a shadow copy is kept on the CPU side. */
+  bool GetCacheMemory() const { return m_bCacheMemory; }
+
+  /*! \brief returns a pointer to the staging texture. */
+  uint8_t* GetPixels() const { return m_pixels; }
+
+  /*! \brief return the size of one row in bytes. */
+  uint32_t GetPitch() const { return GetPitch(m_textureWidth); }
+  /*! \brief return the number of rows. */
+  uint32_t GetRows() const { return GetRows(m_textureHeight); }
+  /*! \brief return the total width of the texture, which might be scaled to fit its displayed size. */
+  uint32_t GetTextureWidth() const { return m_textureWidth; }
+  /*! \brief return the total height of the texture, which might be scaled to fit its displayed size. */
+  uint32_t GetTextureHeight() const { return m_textureHeight; }
+  /*! \brief return the total width of "active" area, which might be equal or lower than the texture width. */
+  uint32_t GetWidth() const { return m_imageWidth; }
+  /*! \brief return the total height of "active" area, which might be equal or lower than the texture height. */
```

Comment:
Why might these be lower than the texture height / width?

Reply:
swscale needs a width with a multiple of 16, but the original texture might not scale to such a size. The skin might request a 500x500 texture, which would be a subset of a 512x512 texture.

A 10x10 texture with a 4x4 block compression scheme would be a subset of a 12x12 texture.

Scanout hardware might need a 256 Byte pitch, so there might be a dead area.

And I'm planning on supporting texture atlases. A 512x512 texture might hold 512 sub textures with a size of 32x32.